py/dml/compat.py

```diff
@@ -178,6 +178,23 @@ class suppress_WLOGMIXUP(CompatFeature):
     last_api_version = api_6
 
 
+@feature
+class lenient_typechecking(CompatFeature):
+    '''This compatibility feature makes DMLC's type checking very inexact and
+    lenient in multiple respects when compared to GCC's type checking of the
+    generated C.
+    This discrepency mostly affects method overrides or uses of `extern`:d C
+    macros, because in those scenarios DMLC can become wholly responsible for
+    proper type checking.
+
+    While migrating from this feature, novel type errors due to uses of
+    `extern`:d macros can often be resolved by changing the signature of
+    the `extern` declaration to more accurately reflect the macro's effective
+    type.
+    '''
```

Comment:
Errgghhh... there are so many different cases that I struggle to think of a **best** example. In fact, the things this affects are so varied that highlighting a common case in particular may be misleading, as it obscures the fact that there are other, _distinct_, cases that affects people's (SSM's) code just as much.
I highlight `extern`:d macros and method overrides because those are the things DMLC are entirely responsible for, but for SSM who just loooove to ignore GCC warnings there are far more and more varied beasts out there.

When it comes to a specific thing that causes most breakage, the most significant is disagreement in const-qualification. But the _ways_ that this causes breakage is varied. For one, there are the fact that `cmp` used to not take const-qualification into account at all; this is what affects method overrides. However there's also the fact that `TPtr.canstore` used to not care about discarding const qualifiers once `void` was involved. This is the big thing (but not the ONLY) thing that is affecting `extern`:d macros (due to the `extern` signature not properly const-qualifying); but once you ignore GCC warnings, the lax logic re. discarding const has more impact than just improper `extern` declarations. It allows three distinct cases, each of which occur in SSM code:
```
const blabla_t *bla = ...;
...
// discarding const qualifier and ignoring GCC warning. 
// DMLC doesn't error as the pointer base type of the parameter is `void`
MM_FREE(bla); 
```

```
// discarding const qualifier and ignoring GCC warning. 
// DMLC doesn't error as the pointer base type of the argument is (`const`) `void`
local some_iface_t *iface = SIM_get_interface(dev, "blablabla");
```

And probably the most promiment:
```
bytes_t whatever;
// discarding const qualifier and ignoring GCC warning. 
// DMLC doesn't error as the pointer base type of the parameter (and also the argument) is (`const`) `void`
memcpy(whatever.data, ..., ...);
```

Disabling `lenient_typechecking` will upgrade each of these cases to a proper error.

in short: _what example should we promote, that actually gives an accurate idea of what people can expect_?

Reply:
Discussed verbally; no example given, instead clarifying the most common kind of type errors you'd observed.